File: app/lua/test/contentAddUser.lua

print('test.contentAddUser')

contentAddUser = {
    --
};

--
local contentsFilePath = './../../test/key/crypto/out/xa_add_user.ctd'; -- contents decrypted
local contentsEncFilePath = './../../test/key/crypto/out/xa_add_user_p1.cte'; -- contents encrypted

-- --
-- local ownerPrikeyFilePath = './../../test/key/crypto/xa_ed_privkey.fin';
-- local ownerPubkeyFilePath = './../../test/key/crypto/xa_ed_pubkey.pem';
-- local superPubkeyFilePath = './../../test/key/crypto/xa_ed_pubkey.pem';

-- --
-- local xaXPrikeyFilePath = './../../test/key/crypto/xa_x_privkey.pem';
-- local xaXPubkeyFilePath = './../../test/key/crypto/xa_x_pubkey.pem';

-- local xbXPrikeyFilePath = './../../test/key/crypto/xb_x_privkey.pem';
-- local xbXPubkeyFilePath = './../../test/key/crypto/xb_x_pubkey.pem';

--
local ownerPrikeyFilePath = './../../../../conf/test/key/ed/key_09/ed_privkey.fin';
local ownerPubkeyFilePath = './../../../../conf/test/key/ed/key_09/ed_pubkey.pem';
local superPubkeyFilePath = './../../../../conf/test/key/ed/key_09/ed_pubkey.pem';

--
local xaXPrikeyFilePath = './../../../../conf/test/key/x25519/09_x_privkey.pem';
local xaXPubkeyFilePath = './../../../../conf/test/key/x25519/09_x_pubkey.pem';

local xbXPrikeyFilePath = './../../../../conf/test/key/x25519/10_x_privkey.pem';
local xbXPubkeyFilePath = './../../../../conf/test/key/x25519/10_x_pubkey.pem';

--
local encJsonFilePath = './../../test/key/crypto/out/xa_add_user_p2.cej'; -- contents encrypted json

--
function contentAddUser.makeContentsJson()
    --
    addUser = AddUser;

    --
    addUser:init();

    -- ownerPrikey
    -- io.write('Owner Private key file path to sign (*.fin) : ');
    -- local ownerPrikeyFilePath = io.read();

    local ownerPrikey = _commonUtil.readBinaryFile(ownerPrikeyFilePath);
    -- local ownerPrikeyHexStr = _commonUtil.bytesToHexStr(ownerPrikey);
    -- print('ownerPrikeyHexStr : ', ownerPrikeyHexStr);

    -- ownerPrikeyPw
    -- io.write('password for *.fin : ');
    -- local ownerPrikeyPw = io.read();
    local ownerPrikeyPw = "asdfQWER1234!@#$";

    -- ownerPubkey
    local ownerPubkey = ed25519GetPubkey(ownerPubkeyFilePath);
    print("ownerPubkey : ", ownerPubkey);
    print("ownerPubkey05 : ", _commonUtil.insertStr(ownerPubkey, '05', 0));
    -- print("str1:sub(1,1) : ", str1:sub(1,1));
    -- print("str1:sub(1+1): ", str1:sub(1+1));

    -- superPubkey
    local superPubkey = ed25519GetPubkey(superPubkeyFilePath);
    print("superPubkey : ", superPubkey);

    -- accountId
    local accountId = 'user_09';

    --
    -- addUser:setMyContentObj('owner_pk', _commonUtil.insertStr(ownerPubkey, '05', 0));
    -- addUser:setMyContentObj('super_pk', _commonUtil.insertStr(superPubkey, '05', 0));
    -- addUser:setMyContentObj('account_id', accountId);
    addUser:setMyContentObj('ownerPubkey', _commonUtil.insertStr(ownerPubkey, '05', 0));
    addUser:setMyContentObj('superPubkey', _commonUtil.insertStr(superPubkey, '05', 0));
    addUser:setMyContentObj('accountId', accountId);
    addUser:setMyContentObj('ownerPrikey', ownerPrikey);
    addUser:setMyContentObj('ownerPrikeyPw', ownerPrikeyPw);

    local plaintext = addUser:jsonStringifyContent(addUser._myContentsName, true);
    -- print('plaintext.len : ', string.len(plaintext));

    -- // for test
    -- local contractJson = json.parse(plaintext);
    -- _commonUtil.prtTable(contractJson['addUser']);
    -- print('ownerPrikeyPw : ', contractJson['addUser']['ownerPrikeyPw']);
    -- ///////////////////////////////////////////////////////////////////////

    _commonUtil.writeBinaryFile(contentsFilePath, plaintext);

    --
    addUser:remove();

    return plaintext;
end

function contentAddUser.makeContentsJsonEnc()
    -- peerXPubkey
    local peerXPubkey;
    -- local peerXPubkeyFilePath = xbXPubkeyFilePath;
    -- peerXPubkey = ed25519GetPubkey(peerXPubkeyFilePath);
    -- peerXPubkey = '308b942a2eae006d4adf46761cfa58b3c86c89d76bd0d05b3ee68b480552ab21'; -- Pubkey of FBN1
    local retCurlMsg = curlHttpGet("http://purichain.com:4000/wallet/key/get/pubkey?x25519", "dummy");
    local culrMsg = json.parse(retCurlMsg);
    -- _commonUtil.prtTable(culrMsg['contents']);
    -- _commonUtil.prtTable(culrMsg);
    --
    peerXPubkey = culrMsg['contents']['xPubkey'];
    print("peerXPubkey : ", peerXPubkey);

    --
    contentsEnc = ContentsEnc;

    --
    contentsEnc:init();

    --
    local myXPrikeyFilePath = xaXPrikeyFilePath;
    local encMsg = contentsEnc:makeContentsJsonEncP1(contentsFilePath, peerXPubkey, myXPrikeyFilePath, contentsEncFilePath);

    --
    local myXPubkeyFilePath = xaXPubkeyFilePath;
    local encJsonMsg = contentsEnc:makeContentsJsonEncP2(encMsg, peerXPubkey, myXPubkeyFilePath, encJsonFilePath)

    --
    contentsEnc:remove();
end

function contentAddUser.makeContentsJsonDec(phase)
    -- peerXPubkey
    local peerXPubkeyFilePath = xaXPubkeyFilePath;
    local peerXPubkey = ed25519GetPubkey(peerXPubkeyFilePath);
    -- print("peerXPubkey : ", peerXPubkey);

    --
    contentsEnc = ContentsEnc;

    --
    contentsEnc:init();

    --
    local myXPrikeyFilePath = xbXPrikeyFilePath;
    
    if (phase == 1) then
        -- Phase 1
        local plaintext = contentsEnc:makeContentsJsonDec(nil, contentsEncFilePath, peerXPubkey, myXPrikeyFilePath);
    else
        -- Phase 2
        local plaintext = contentsEnc:makeContentsJsonDec(encJsonFilePath, nil, peerXPubkey, myXPrikeyFilePath);
    end

    --
    contentsEnc:remove();

    return plaintext;
end

function contentAddUser.sendContents(filePath)
    local plaintext = _commonUtil.readBinaryFile(filePath);
    -- print('plaintext.len : ', string.len(plaintext));
    print('plaintext.len : ', string.len(plaintext));
    print('plaintext : ', plaintext);

    -- // for test
    -- local contractJson = json.parse(plaintext);
    -- _commonUtil.prtTable(contractJson['addUser']);
    -- print('ownerPrikey4 : ', contractJson['addUser']['ownerPrikey']);
    -- print('ownerPrikeyPw : ', contractJson['addUser']['ownerPrikeyPw']);

    -- local ownerPrikeyHexStr = _commonUtil.bytesToHexStr(contractJson['addUser']['ownerPrikey']);
    -- print('ownerPrikeyHexStr : ', ownerPrikeyHexStr);
    -- ///////////////////////////////////////////////////////////////////////

    local plaintextHexStr = _commonUtil.bytesToHexStr(plaintext);
    print('plaintextHexStr.len : ', string.len(plaintextHexStr));
    print('plaintextHexStr : ', plaintextHexStr);

    -- local myContents = "contents=" .. plaintext;
    local myContents = "contents=" .. plaintextHexStr;

    curlHttpPost("http://purichain.com:4000/contract/tool/json", myContents);
end

function contentAddUser.sendContentsEnc(filePath)
    local encJsonMsg = _commonUtil.readBinaryFile(filePath);
    print('encJsonMsg : ', encJsonMsg);

    local myContentsEnc = "contentsEnc=" .. encJsonMsg;

    curlHttpPost("http://purichain.com:4000/contract/tool/json", myContentsEnc);
end

function contentAddUser.testCase()
    --
    local rtnJsonStr = contentAddUser.makeContentsJson();
    -- contentAddUser.sendContents(contentsFilePath);

    --
    local encMsg = contentAddUser.makeContentsJsonEnc();

    --
    contentAddUser.makeContentsJsonDec(2);

    contentAddUser.sendContentsEnc(encJsonFilePath);
end

File: app/nodejs/test/contentAddUser.js
//
const fs = require('fs');

//
const cryptoSsl = require("./../linux/build/Release/crypto-ssl");

//
const config = require("./../njsConn/config/config.js");
const define = require("./../njsConn/config/define.js");

const contentsEnc = require("../njsConn/core/contents/contentsEnc.js");

//
let contentsFilePath = './../../test/key/crypto/out/xa_add_user.ctd'; // contents decrypted
let contentsEncFilePath = './../../test/key/crypto/out/xa_add_user_p1.cte'; // contents encrypted

//
let ownerPrikeyFilePath = './../../../../conf/test/key/ed/key_09/ed_privkey.fin';
let ownerPubkeyFilePath = './../../../../conf/test/key/ed/key_09/ed_pubkey.pem';
let superPubkeyFilePath = './../../../../conf/test/key/ed/key_09/ed_pubkey.pem';

//
let xaXPrikeyFilePath = './../../../../conf/test/key/x25519/09_x_privkey.pem';
let xaXPubkeyFilePath = './../../../../conf/test/key/x25519/09_x_pubkey.pem';
let xbXPrikeyFilePath = './../../../../conf/test/key/x25519/10_x_privkey.pem';
let xbXPubkeyFilePath = './../../../../conf/test/key/x25519/10_x_pubkey.pem';

//
let encJsonFilePath = './../../test/key/crypto/out/xa_add_user_p2.cej'; // contents encrypted json

//
module.exports.makeContentsJson = async () => {
    ////////////////////////////////////////////////////////////////////
    // ownerPrikey
    let ownerPrikey = fs.readFileSync(ownerPrikeyFilePath, 'binary');

    // let ownerPrikeyHexStr = Buffer.from(ownerPrikey,'binary').toString('hex');
    // console.log("ownerPrikeyHexStr : " + ownerPrikeyHexStr);

    // ownerPrikeyPw
    let ownerPrikeyPw = "asdfQWER1234!@#$";

    // let stdin = process.openStdin();
    // console.log('password for *.fin : ');
    // stdin.addListener("data", function(d) {
    //     // note:  d is an object, and when converted to a string it will
    //     // end with a linefeed.  so we (rather crudely) account for that  
    //     // with toString() and then trim() 
    //     console.log("you entered: [" +  d.toString().trim() + "]");

    //     ownerPrikeyPw = d.toString().trim();
    // });

    // ownerPubkey
    let ownerPubkey = cryptoSsl.ed25519GetPubkey(ownerPubkeyFilePath);
    console.log("ownerPubkey : " + ownerPubkey);

    // superPubkey
    let superPubkey = cryptoSsl.ed25519GetPubkey(superPubkeyFilePath);
    console.log("superPubkey : " + superPubkey);

    // accountId
    let accountId = 'user_01';

    //
    let contentsJson = {
        addUser : {
            ownerPrikey : ownerPrikey,
            ownerPrikeyPw : ownerPrikeyPw,
            ownerPubkey : define.CONTRACT_DEFINE.ED_PUB_IDX + ownerPubkey,
            superPubkey : define.CONTRACT_DEFINE.ED_PUB_IDX + superPubkey,
            accountId : accountId
        }
    };

    let plaintext = JSON.stringify(contentsJson);
    // console.log("plaintext.length : " + plaintext.length);
    // console.log("plaintext : " + plaintext);

    fs.writeFileSync(contentsFilePath, plaintext, 'binary');
    // fs.writeFileSync(contentsFilePath, plaintext);

    return plaintext;
}

//
module.exports.makeContentsJsonEnc = async () => {
    ////////////////////////////////////////////////////////////////////
    // Phase 1
    // plaintext
    // let plaintext = fs.readFileSync(contentsFilePath, 'binary');
    let plaintext = fs.readFileSync(contentsFilePath);
    // console.log("plaintext : " + plaintext);

    //
    let plaintextHexStr = Buffer.from(plaintext, 'utf-8').toString('hex');

    // console.log("plaintextHexStr.length : " + plaintextHexStr.length);
    // console.log("plaintextHexStr : " + plaintextHexStr);

    ////////////////////////////////////////////////////////////////////
    // peerXPubkey
    let peerXPubkey;
    // peerXPubkey = cryptoSsl.ed25519GetPubkey(xbXPubkeyFilePath);
    // peerXPubkey = '308b942a2eae006d4adf46761cfa58b3c86c89d76bd0d05b3ee68b480552ab21'; // Pubkey of FBN1
    let retCurlMsg = cryptoSsl.curlHttpGet("http://purichain.com:4000/wallet/key/get/pubkey?x25519", "dummy");
    let culrMsg = JSON.parse(retCurlMsg);
    //
    peerXPubkey = culrMsg.contents.xPubkey;;

    console.log("peerXPubkey : " + peerXPubkey);
    
    // myPrikeyFile
    // let myXPrikeyFile = fs.readFileSync(xaXPrikeyFilePath, 'binary');
    let myXPrikeyFile = fs.readFileSync(xaXPrikeyFilePath);
    // console.log("myXPrikeyFile.length : " + myXPrikeyFile.length);
    // console.log("myXPrikeyFile : " + myXPrikeyFile);

    ////////////////////////////////////////////////////////////////////
    // encMsg
    let encMsg = cryptoSsl.x25519MixEnc(myXPrikeyFile, peerXPubkey, plaintextHexStr, plaintextHexStr.length);
    console.log("encMsg : " + encMsg);

    let encMsg_b = Buffer.from(encMsg, 'hex');
    fs.writeFileSync(contentsEncFilePath, encMsg_b, 'binary');

    ////////////////////////////////////////////////////////////////////
    // Phase 2
    // myXPubkey
    let myXPubkey = cryptoSsl.ed25519GetPubkey(xaXPubkeyFilePath);
    // console.log("myXPubkey : " + myXPubkey);

    //
    let encXPubkeyF = define.CONTRACT_DEFINE.ED_PUB_IDX + peerXPubkey;
    let myXPubkeyF = define.CONTRACT_DEFINE.ED_PUB_IDX + myXPubkey;
    let encJsonMsg = contentsEnc.makeEncJsonMsg(encMsg, encXPubkeyF, myXPubkeyF, encJsonFilePath);

    //
    return encJsonMsg;
}

//
module.exports.makeContentsJsonDec = async (phase) => {
    ////////////////////////////////////////////////////////////////////
    let encMsgHexStr;

    if (phase === 2)
    {
        let encJsonMsg = fs.readFileSync(encJsonFilePath);
        // console.log("encJsonMsg.length : " + encJsonMsg.length);
    
        //
        let encJson = JSON.parse(encJsonMsg);

        encMsgHexStr = encJson.jsonEnc.contentsEnc;
    }
    else // phase === 1
    {
        // encMsg
        let encMsg;

        // encMsg = fs.readFileSync(contentsFilePath, 'binary');
        encMsg = fs.readFileSync(contentsEncFilePath);

        // console.log("encMsg : " + encMsg);

        encMsgHexStr = Buffer.from(encMsg, 'utf-8').toString('hex');
    }

    // console.log("encMsgHexStr.length : " + encMsgHexStr.length);
    // console.log("encMsgHexStr : " + encMsgHexStr);

    ////////////////////////////////////////////////////////////////////
    // peerXPubkey
    let peerXPubkey = cryptoSsl.ed25519GetPubkey(xaXPubkeyFilePath);
    // console.log("peerXPubkey : " + peerXPubkey);

    // myXPrikey
    // let myXPrikeyFile = fs.readFileSync(xbXPrikeyFilePath, 'binary');
    let myXPrikeyFile = fs.readFileSync(xbXPrikeyFilePath);
    // console.log("myXPrikeyFile.length : " + myXPrikeyFile.length);
    // console.log("myXPrikeyFile : " + myXPrikeyFile);

    plaintext = cryptoSsl.x25519MixDec(myXPrikeyFile, peerXPubkey, encMsgHexStr, encMsgHexStr.length);
    // console.log("plaintext.length : " + plaintext.length);
    // console.log("plaintext : " + plaintext);

    return plaintext;
}

module.exports.sendContents = async (filePath) => {
    let plaintext = fs.readFileSync(filePath);
    console.log("plaintext : " + plaintext);

    //
    let plaintextHexStr = Buffer.from(plaintext, 'binary').toString('hex');
    console.log('plaintextHexStr : ' + plaintextHexStr);

    // // for test
    // let plaintext2 = Buffer.from(plaintextHexStr, 'hex');
    // let contractJson = JSON.parse(plaintext2.toString('binary'));
    // console.log('ownerPrikeyPw : ' + contractJson.addUser.ownerPrikeyPw);

    // fs.writeFileSync('./../../test/key/crypto/out/xa_add_user2.ctd', contractJson.addUser.ownerPrikey, 'binary');
    // // ///////////////////////////////////////////////////////////////////////

    let myContents = "contents=" + plaintextHexStr;

    cryptoSsl.curlHttpPost("http://purichain.com:4000/contract/tool/json", myContents);
}

module.exports.sendContentsEnc = async (filePath) => {
    let encJsonMsg = fs.readFileSync(filePath);
    console.log('encJsonMsg : ' + encJsonMsg);

    let myContentsEnc = "contentsEnc=" + encJsonMsg;

    cryptoSsl.curlHttpPost("http://purichain.com:4000/contract/tool/json", myContentsEnc);
}

module.exports.testCase = async () => {
    this.makeContentsJson();
    // this.sendContents(contentsFilePath);
    this.makeContentsJsonEnc();
    // this.makeContentsJsonDec(2);
    this.sendContentsEnc(encJsonFilePath);
}
